algorithms/sorting/pdqsort.py

```diff
@@ -0,0 +1,177 @@
+
+    Parameters:
+    arr (List[int]): The list to sort.
+    low (int): The starting index of the portion to sort.
+    high (int): The ending index of the portion to sort.
+    depth_limit (int): The maximum recursion depth.
+    """
+    while low < high:
+        # Use insertion sort for small partitions
+        if high - low <= 16:
+            insertion_sort(arr, low, high)
+            return
+
+        if depth_limit == 0:
+            # Switch to heapsort if depth limit is reached
+            heap_sort(arr, low, high)
+            return
+
+        pivot_index = partition(arr, low, high)
+        pdqsort_recursive(arr, low, pivot_index - 1, depth_limit - 1)
+        # Tail recursion elimination
+        low = pivot_index + 1
+
+
+def pdqsort(arr):
+    """
+    Sorts the entire array using the Pattern-Defeating Quicksort (PDQSort) algorithm.
+
+    PDQSort is an optimized version of Quicksort that includes enhancements to deal with
+    various patterns in the input data, such as already sorted or nearly sorted sequences.
+    It dynamically switches between different sorting strategies based on the characteristics
+    of the data and recursion depth to achieve better performance.
+
+    PDQSort employs the following techniques:
+    - Insertion Sort for small arrays to reduce overhead.
+    - Partitioning around a pivot to divide the array into subarrays.
+    - Heap Sort as a fallback when the recursion depth limit is reached, preventing worst-case
+      scenarios that could degrade performance to O(n^2).
+    - Tail call optimization to avoid excessive recursive calls and stack overflow.
+
+    Parameters:
+    arr (List[int]): The list of integers to sort. The list is sorted in place.
+
+    Returns:
+    None: The function modifies the input list in place and does not return a value.
+
+    Example:
+    >>> arr = [3, 6, 8, 10, 1, 2, 1]
+    >>> pdqsort(arr)
+    >>> print(arr)
+    [1, 1, 2, 3, 6, 8, 10]
+    """
+    max_depth = (len(arr).bit_length() - 1) * 2
+    pdqsort_recursive(arr, 0, len(arr) - 1, max_depth)
+
+
+class TestPDQSort(unittest.TestCase):
+    def test_empty_list(self):
+        arr = []
+        pdqsort(arr)
+        self.assertEqual(arr, [])
+
+    def test_sorted_list(self):
+        arr = [1, 2, 3, 4, 5]
+        pdqsort(arr)
+        self.assertEqual(arr, [1, 2, 3, 4, 5])
+
+    def test_reverse_sorted_list(self):
+        arr = [5, 4, 3, 2, 1]
+        pdqsort(arr)
+        self.assertEqual(arr, [1, 2, 3, 4, 5])
+
+    def test_random_list(self):
+        arr = [5, 2, 8, 3, 1, 9, 4, 6, 7]
+        pdqsort(arr)
+        self.assertEqual(arr, [1, 2, 3, 4, 5, 6, 7, 8, 9])
+
+    def test_duplicates(self):
+        arr = [3, 2, 1, 3, 1, 2]
+        pdqsort(arr)
+        self.assertEqual(arr, [1, 1, 2, 2, 3, 3])
+
+    def test_large_random_list(self):
+        arr = random.sample(range(1000000), 1000)
+        sorted_arr = sorted(arr)
+        pdqsort(arr)
+        self.assertEqual(arr, sorted_arr)
+
+
```

Comment:
We should add tests for list of strings like this:
```suggestion
    def test_string_list(self):
        arr = ["banana", "apple", "orange", "grape", "kiwi"]
        pdqsort(arr)
        self.assertEqual(arr, ["apple", "banana", "grape", "kiwi", "orange"])

    def test_empty_string_list(self):
        arr = ["", "apple", "banana", "", "orange", "", "grape", "kiwi", ""]
        pdqsort(arr)
        self.assertEqual(
            arr, ["", "", "", "", "apple", "banana", "grape", "kiwi", "orange"]
        )
        
        
```